### DISCUSSION TITLE
condition order counterbalance between participants

### DISCUSSION THREAD

**Discussion by ricmo22** [jspsych/jsPsych#1532] [Q&A]
Hi!
My study has two conditions: 1 and 2. I want half of the participants to do condition 1 first, then condition 2; while the other half do condition 2 before condition 1. 
Additionally, each condition has its own instruction and practice. So, this study has instruction 1, practice 1, main test 1, instruction 2, practice 2 and main test 2. 
Practice and condition are using different stimuli. 
Is there a way for me to group up each instruction + practice + main test together in each condition? Can I also balance the order between participants?
I explore the shuffle function, but not sure if I am using it right.

Please help! many thanks!
Ric

**Comment by JAQuent** [comment]
Couldn't  you simply add an if-statement where if 1 you push condition 1 then condition 2 (along the respective instructions) to the timeline and if 2 you do it the other way around?

**Reply by becky-gilbert** [reply]
I agree that you could do this with an `if` statement. Here's an example:
```js
var timeline = [];
if (condition == 1) {
  timeline.push(instruction1, practice1, maintest1, instruction2, practice2, maintest2);
} else if (condition == 2) {
  timeline.push(instruction2, practice2, maintest2, instruction1, practice1, maintest1);
}
```
Note that you can't actually counterbalance conditions with jsPsych, because that requires knowing something about the history of conditions that have been previously assigned (e.g. the last condition that was assigned, or the number of assignments to each condition). So to do condition counterbalancing, you need to access permanently stored data from a server. In jsPsych, you can randomize the condition assignment.
```js
var condition = jsPsych.randomization.sampleWithoutReplacement([1,2],1)[0];
```
See the [randomization documentation](https://www.jspsych.org/core_library/jspsych-randomization/#jspsychrandomizationsamplewithoutreplacement) for more info on the `sampleWithoutReplacement` function, and the [Running Experiments page](https://www.jspsych.org/overview/running-experiments/#permanent-data-storage) for more info on counterbalancing and permanent data storage.

**Reply by jodeleeuw** [reply]
Just to add to Becky's note about counterbalancing: cognition.run can do this for you automatically by providing a URL variable that assigns conditions. I also have [shared code to do this using a MySQL/PHP backend](https://github.com/vassar-cogscilab/php-mysql-save-data).
